From 003522302c9a330ad7040aa56adbf47b97d98b79 Mon Sep 17 00:00:00 2001
From: SamulKyull <SamulKyull@SamulKyull.xyz>
Date: Sun, 23 Nov 2025 20:02:14 +0800
Subject: [PATCH] refactor(sdhci): update sdhci implementation to use new
 sunxi_sdhci_t structure

Replace old sdhci_t with new sunxi_sdhci_t structure for better hardware abstraction
Add support for eMMC controller (sdhci2) alongside existing SD card controller
Update board configuration to enable MMC_V2 support
Remove unused DRAM payload configuration
---
 board/mcore-r818/board.c              | 92 +++++++++++++++++++++------
 board/mcore-r818/extlinux_boot/main.c |  2 +-
 board/mcore-r818/smhc_test/main.c     | 11 ++--
 board/mcore-r818/syter_boot/main.c    |  2 +-
 cmake/board/mcore-r818.cmake          | 31 ++-------
 5 files changed, 86 insertions(+), 52 deletions(-)

diff --git a/board/mcore-r818/board.c b/board/mcore-r818/board.c
index e0d53853..6b00f5cf 100644
--- a/board/mcore-r818/board.c
+++ b/board/mcore-r818/board.c
@@ -14,10 +14,11 @@
 
 #include <mmu.h>
 
+#include <mmc/sys-sdhci.h>
+
 #include <sys-dram.h>
 #include <sys-gpio.h>
 #include <sys-i2c.h>
-#include <sys-sdcard.h>
 #include <sys-sid.h>
 #include <sys-spi.h>
 #include <sys-uart.h>
@@ -44,23 +45,78 @@ sunxi_serial_t uart_dbg = {
 				},
 };
 
-sdhci_t sdhci0 = {
-		.name = "sdhci0",
-		.id = 0,
-		.reg = (sdhci_reg_t *) SUNXI_SMHC0_BASE,
-		.voltage = MMC_VDD_27_36,
-		.width = MMC_BUS_WIDTH_4,
-		.clock = MMC_CLK_50M,
-		.removable = 0,
-		.isspi = FALSE,
-		.skew_auto_mode = FALSE,
-		.sdhci_pll = CCU_MMC_CTRL_PLL_PERIPH1X,
-		.gpio_clk = {GPIO_PIN(GPIO_PORTF, 2), GPIO_PERIPH_MUX2},
-		.gpio_cmd = {GPIO_PIN(GPIO_PORTF, 3), GPIO_PERIPH_MUX2},
-		.gpio_d0 = {GPIO_PIN(GPIO_PORTF, 1), GPIO_PERIPH_MUX2},
-		.gpio_d1 = {GPIO_PIN(GPIO_PORTF, 0), GPIO_PERIPH_MUX2},
-		.gpio_d2 = {GPIO_PIN(GPIO_PORTF, 5), GPIO_PERIPH_MUX2},
-		.gpio_d3 = {GPIO_PIN(GPIO_PORTF, 4), GPIO_PERIPH_MUX2},
+sunxi_sdhci_t sdhci0 = {
+		.name = "SD Card",
+		.id = MMC_CONTROLLER_0,
+		.reg_base = SUNXI_SMHC0_BASE,
+		.sdhci_mmc_type = MMC_TYPE_SD,
+		.max_clk = 50000000,
+		.width = SMHC_WIDTH_4BIT,
+		.dma_des_addr = SDRAM_BASE + 0x10080000,
+		.pinctrl =
+				{
+						.gpio_clk = {GPIO_PIN(GPIO_PORTF, 2), GPIO_PERIPH_MUX2},
+						.gpio_cmd = {GPIO_PIN(GPIO_PORTF, 3), GPIO_PERIPH_MUX2},
+						.gpio_d0 = {GPIO_PIN(GPIO_PORTF, 1), GPIO_PERIPH_MUX2},
+						.gpio_d1 = {GPIO_PIN(GPIO_PORTF, 0), GPIO_PERIPH_MUX2},
+						.gpio_d2 = {GPIO_PIN(GPIO_PORTF, 5), GPIO_PERIPH_MUX2},
+						.gpio_d3 = {GPIO_PIN(GPIO_PORTF, 4), GPIO_PERIPH_MUX2},
+				},
+		.clk_ctrl =
+				{
+						.gate_reg_base = CCU_BASE + CCU_SMHC_BGR_REG,
+						.gate_reg_offset = SDHCI_DEFAULT_CLK_GATE_OFFSET(0),
+						.rst_reg_base = CCU_BASE + CCU_SMHC_BGR_REG,
+						.rst_reg_offset = SDHCI_DEFAULT_CLK_RST_OFFSET(0),
+				},
+		.sdhci_clk =
+				{
+						.reg_base = CCU_BASE + CCU_SMHC0_CLK_REG,
+						.reg_factor_n_offset = SDHCI_DEFAULT_CLK_FACTOR_N_OFFSET,
+						.reg_factor_m_offset = SDHCI_DEFAULT_CLK_FACTOR_M_OFFSET,
+						.clk_sel = 0x1,
+						.parent_clk = 300000000,
+				},
+};
+
+sunxi_sdhci_t sdhci2 = {
+		.name = "eMMC",
+		.id = MMC_CONTROLLER_2,
+		.reg_base = SUNXI_SMHC2_BASE,
+		.sdhci_mmc_type = MMC_TYPE_EMMC,
+		.max_clk = 50000000,
+		.width = SMHC_WIDTH_8BIT,
+		.dma_des_addr = SDRAM_BASE + 0x10880000,
+		.pinctrl =
+				{
+						.gpio_clk = {GPIO_PIN(GPIO_PORTC, 5), GPIO_PERIPH_MUX3},
+						.gpio_cmd = {GPIO_PIN(GPIO_PORTC, 6), GPIO_PERIPH_MUX3},
+						.gpio_d0 = {GPIO_PIN(GPIO_PORTC, 10), GPIO_PERIPH_MUX3},
+						.gpio_d1 = {GPIO_PIN(GPIO_PORTC, 13), GPIO_PERIPH_MUX3},
+						.gpio_d2 = {GPIO_PIN(GPIO_PORTC, 15), GPIO_PERIPH_MUX3},
+						.gpio_d3 = {GPIO_PIN(GPIO_PORTC, 8), GPIO_PERIPH_MUX3},
+						.gpio_d4 = {GPIO_PIN(GPIO_PORTC, 9), GPIO_PERIPH_MUX3},
+						.gpio_d5 = {GPIO_PIN(GPIO_PORTC, 11), GPIO_PERIPH_MUX3},
+						.gpio_d6 = {GPIO_PIN(GPIO_PORTC, 14), GPIO_PERIPH_MUX3},
+						.gpio_d7 = {GPIO_PIN(GPIO_PORTC, 16), GPIO_PERIPH_MUX3},
+						.gpio_ds = {GPIO_PIN(GPIO_PORTC, 0), GPIO_PERIPH_MUX3},
+						.gpio_rst = {GPIO_PIN(GPIO_PORTC, 1), GPIO_PERIPH_MUX3},
+				},
+		.clk_ctrl =
+				{
+						.gate_reg_base = CCU_BASE + CCU_SMHC_BGR_REG,
+						.gate_reg_offset = SDHCI_DEFAULT_CLK_GATE_OFFSET(2),
+						.rst_reg_base = CCU_BASE + CCU_SMHC_BGR_REG,
+						.rst_reg_offset = SDHCI_DEFAULT_CLK_RST_OFFSET(2),
+				},
+		.sdhci_clk =
+				{
+						.reg_base = CCU_BASE + CCU_SMHC2_CLK_REG,
+						.reg_factor_n_offset = SDHCI_DEFAULT_CLK_FACTOR_N_OFFSET,
+						.reg_factor_m_offset = SDHCI_DEFAULT_CLK_FACTOR_M_OFFSET,
+						.clk_sel = 0x1,
+						.parent_clk = 400000000,
+				},
 };
 
 sunxi_i2c_t i2c_pmu = {
diff --git a/board/mcore-r818/extlinux_boot/main.c b/board/mcore-r818/extlinux_boot/main.c
index c351c8ed..7c50a4a7 100644
--- a/board/mcore-r818/extlinux_boot/main.c
+++ b/board/mcore-r818/extlinux_boot/main.c
@@ -53,7 +53,7 @@ extern sunxi_serial_t uart_dbg;
 
 extern sunxi_i2c_t i2c_pmu;
 
-extern sdhci_t sdhci0;
+extern sunxi_sdhci_t sdhci0;
 
 extern uint32_t dram_para[32];
 
diff --git a/board/mcore-r818/smhc_test/main.c b/board/mcore-r818/smhc_test/main.c
index 893f9968..65a7cb6d 100644
--- a/board/mcore-r818/smhc_test/main.c
+++ b/board/mcore-r818/smhc_test/main.c
@@ -15,13 +15,14 @@
 #include <sys-gpio.h>
 #include <sys-i2c.h>
 #include <sys-sid.h>
+#include <mmc/sys-sdhci.h>
 #include <sys-sdcard.h>
 #include <sys-spi.h>
 #include <sys-uart.h>
 
 extern sunxi_serial_t uart_dbg;
 extern uint32_t dram_para[32];
-extern sdhci_t sdhci0;
+extern sunxi_sdhci_t sdhci2;
 extern sunxi_i2c_t i2c_pmu;
 
 static void set_pmu_fin_voltage(char *power_name, uint32_t voltage) {
@@ -58,14 +59,14 @@ int main(void) {
 	sunxi_clk_dump();
 
 	/* Initialize the SD host controller. */
-	if (sunxi_sdhci_init(&sdhci0) != 0) {
-		printk_error("SMHC: %s controller init failed\n", sdhci0.name);
+	if (sunxi_sdhci_init(&sdhci2) != 0) {
+		printk_error("SMHC: %s controller init failed\n", sdhci2.name);
 	} else {
-		printk_info("SMHC: %s controller initialized\n", sdhci0.name);
+		printk_info("SMHC: %s controller initialized\n", sdhci2.name);
 	}
 
 	/* Initialize the SD card and check if initialization is successful. */
-	if (sdmmc_init(&card0, &sdhci0) != 0) {
+	if (sdmmc_init(&card0, &sdhci2) != 0) {
 		printk_warning("SMHC: init failed\n");
 	}
 
diff --git a/board/mcore-r818/syter_boot/main.c b/board/mcore-r818/syter_boot/main.c
index e4cf4098..4da961c7 100644
--- a/board/mcore-r818/syter_boot/main.c
+++ b/board/mcore-r818/syter_boot/main.c
@@ -51,7 +51,7 @@
 
 extern sunxi_serial_t uart_dbg;
 extern sunxi_i2c_t i2c_pmu;
-extern sdhci_t sdhci0;
+extern sunxi_sdhci_t sdhci0;
 extern uint32_t dram_para[32];
 
 extern int ar100s_gpu_fix(void);
diff --git a/cmake/board/mcore-r818.cmake b/cmake/board/mcore-r818.cmake
index 85645903..f0664e7f 100644
--- a/cmake/board/mcore-r818.cmake
+++ b/cmake/board/mcore-r818.cmake
@@ -5,19 +5,16 @@ set(CONFIG_ARCH_ARM32_ARM64 True)
 set(CONFIG_CHIP_SUN50IW10 True)
 set(CONFIG_CHIP_WITHPMU True)
 set(CONFIG_CHIP_GPIO_V1 True)
+set(CONFIG_CHIP_MMC_V2 True)
 set(CONFIG_CHIP_MINSTACK False)
 
 set(CONFIG_BOARD_MCORE-R818 True)
 
 add_definitions(-DCONFIG_CHIP_GPIO_V1)
 add_definitions(-DCONFIG_CHIP_SUN50IW10)
-
-set(CONFIG_USE_DRAM_PAYLOAD True)
-set(CONFIG_USE_PREBUILT_DRAM_PAYLOAD True)
-set(CONFIG_USE_DRAM_PAYLOAD_SOURCE_PATH "${CMAKE_SOURCE_DIR}/payloads/sun50iw10_libdram")
-set(CONFIG_USE_DRAM_PAYLOAD_BIN_PATH "${CONFIG_USE_DRAM_PAYLOAD_SOURCE_PATH}/output/ddr.bin")
-set(CONFIG_USE_DRAM_PAYLOAD_FILE_PATH "${CMAKE_SOURCE_DIR}/board/100ask-ros/payloads/init_dram_bin.c")
-set(CONFIG_USE_DRAM_PAYLOAD_SECTION "init_dram_bin")
+add_definitions(-DCONFIG_CHIP_MMC_V2)
+add_definitions(-DCONFIG_FATFS_CACHE_SIZE=0xFFFFFFF)
+add_definitions(-DCONFIG_FATFS_CACHE_ADDR=0x4400000)
 
 # Set the cross-compile toolchain
 if(DEFINED ENV{LINARO_GCC_721_PATH})
@@ -47,23 +44,3 @@ set(ARCH_BIN_SRAM_LENGTH "128K")
 
 set(ARCH_FEL_START_ADDRESS "0x00028000")
 set(ARCH_FEL_SRAM_LENGTH "128K")
-
-if(NOT CONFIG_USE_PREBUILT_DRAM_PAYLOAD)
-    # Create an external project and build it
-    ExternalProject_Add(
-        init_dram
-        PREFIX init_dram
-        SOURCE_DIR "${CONFIG_USE_DRAM_PAYLOAD_SOURCE_PATH}"
-        INSTALL_COMMAND ""
-        CONFIGURE_COMMAND ""
-        BUILD_COMMAND make -C ${CONFIG_USE_DRAM_PAYLOAD_SOURCE_PATH}
-        BUILD_IN_SOURCE 1
-    )
-
-    # Create inital init dram bin file for build
-    add_custom_command(
-        TARGET init_dram
-        POST_BUILD COMMAND ${CMAKE_BIN2ARRAY} ${CONFIG_USE_DRAM_PAYLOAD_BIN_PATH} ${CONFIG_USE_DRAM_PAYLOAD_FILE_PATH} ${CONFIG_USE_DRAM_PAYLOAD_SECTION}
-        COMMENT "Generate DRAM LIB Payload ${CONFIG_USE_DRAM_PAYLOAD_BIN_PATH} for ${CONFIG_USE_DRAM_PAYLOAD_FILE_PATH}"
-    )
-endif()
\ No newline at end of file
-- 
2.34.1

